### PR TITLE
UCT/SM/KNEM: Mark knem regions DONTFORK

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by yosefe** [comment]
bot:pipe:retest

File: src/uct/sm/scopy/knem/knem_md.c [hunk 1]
```diff
@@ -121,6 +124,23 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
...
+    page_size = ucs_get_page_size();
+    start     = ucs_align_down((intptr_t)address, page_size);
+    end       = ucs_align_up((intptr_t)address + length, page_size);
+    rc        = madvise((void *)start, end - start, MADV_DONTFORK);
+    if (rc == EINVAL) {
+        ucs_get_mem_page_size(address, length, &unused, &page_size);
+        start = ucs_align_down((intptr_t)address, page_size);
+        end   = ucs_align_up((intptr_t)address + length, page_size);
+        rc    = madvise((void *)start, end - start, MADV_DONTFORK);
+        if (rc) {
```
Comment (dmitrygx) [inline src/uct/sm/scopy/knem/knem_md.c:136]: [minor] `rc != 0`

File: src/uct/sm/scopy/knem/knem_md.c [hunk 2]
```diff
@@ -121,6 +124,23 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
...
+            if (!silent) {
+                ucs_error("KNEM madvice DONTFORK failed: %m");
```
Comment (dmitrygx) [inline src/uct/sm/scopy/knem/knem_md.c:138]: [minor] print `address` and `length` values?

File: src/uct/sm/scopy/knem/knem_md.c [hunk 3]
```diff
@@ -121,6 +124,23 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
     create.flags = 0;
     create.protection = PROT_READ | PROT_WRITE;
 
+    page_size = ucs_get_page_size();
+    start     = ucs_align_down((intptr_t)address, page_size);
+    end       = ucs_align_up((intptr_t)address + length, page_size);
+    rc        = madvise((void *)start, end - start, MADV_DONTFORK);
```
Comment (dmitrygx) [inline src/uct/sm/scopy/knem/knem_md.c:130]: [minor] `(void*)`

File: src/uct/sm/scopy/knem/knem_md.c [hunk 4]
```diff
@@ -121,6 +124,23 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
...
+    if (rc == EINVAL) {
+        ucs_get_mem_page_size(address, length, &unused, &page_size);
+        start = ucs_align_down((intptr_t)address, page_size);
+        end   = ucs_align_up((intptr_t)address + length, page_size);
+        rc    = madvise((void *)start, end - start, MADV_DONTFORK);
```
Comment (dmitrygx) [inline src/uct/sm/scopy/knem/knem_md.c:135]: [minor] `(void*)`

File: src/uct/sm/scopy/knem/knem_md.c [hunk 5]
```diff
@@ -121,6 +124,23 @@ static ucs_status_t uct_knem_mem_reg_internal(uct_md_h md, void *address, size_t
     create.flags = 0;
     create.protection = PROT_READ | PROT_WRITE;
 
+    page_size = ucs_get_page_size();
+    start     = ucs_align_down((intptr_t)address, page_size);
```
Comment (dmitrygx) [inline src/uct/sm/scopy/knem/knem_md.c:128]: maybe add wrapper?
```
int uct_knem_mem_advise_dont_fork(const void *address, size_t length, size_t page_size)
{
    intptr_t start = ucs_align_down((intptr_t)address, page_size);
    intptr_t end   = ucs_align_up((intptr_t)address + length, page_size);
    return madvise((void *)start, end - start, MADV_DONTFORK);
}
```
wdyt?

**Pull request closed without merging**